git_machete/cmd.py

```diff
@@ -1094,20 +1102,81 @@ def merge_base(cli_ctxt: CommandLineContext, sha1: str, sha2: str) -> str:
 
 # Note: the 'git rev-parse --verify' validation is not performed in case for either of earlier/later
 # if the corresponding prefix is empty AND the revision is a 40 hex digit hash.
-def is_ancestor(cli_ctxt: CommandLineContext, earlier_revision: str, later_revision: str, earlier_prefix: str = "refs/heads/", later_prefix: str = "refs/heads/") -> bool:
-    if earlier_prefix == "" and is_full_sha(earlier_revision):
-        earlier_sha: Optional[str] = earlier_revision
-    else:
-        earlier_sha = commit_sha_by_revision(cli_ctxt, earlier_revision, earlier_prefix)
-    if later_prefix == "" and is_full_sha(later_revision):
-        later_sha: Optional[str] = later_revision
-    else:
-        later_sha = commit_sha_by_revision(cli_ctxt, later_revision, later_prefix)
+def is_ancestor(
+    cli_ctxt: CommandLineContext,
+    earlier_revision: str,
+    later_revision: str,
+    earlier_prefix: str = "refs/heads/",
+    later_prefix: str = "refs/heads/",
+) -> bool:
+    earlier_sha = full_sha(cli_ctxt, earlier_revision, earlier_prefix)
+    later_sha = full_sha(cli_ctxt, later_revision, later_prefix)
+
     if earlier_sha == later_sha:
         return True
     return merge_base(cli_ctxt, earlier_sha, later_sha) == earlier_sha
 
 
+# Determine if any revisions in later_revision contain a tree with identical
+# contents to earlier_revision, indicating that later_revision contains a
+# rebase or squash merge of earlier_revision.
+def contains_equivalent_tree(
+    cli_ctxt: CommandLineContext,
+    earlier_revision: str,
+    later_revision: str,
+    earlier_prefix: str = "refs/heads/",
+    later_prefix: str = "refs/heads/",
+) -> bool:
+    earlier_sha = full_sha(cli_ctxt, earlier_revision, earlier_prefix)
+    later_sha = full_sha(cli_ctxt, later_revision, later_prefix)
+
+    if earlier_sha == later_sha:
+        return True
+
+    debug(
+        cli_ctxt,
+        "contains_equivalent_tree",
+        f"earlier_revision={earlier_revision} later_revision={later_revision}",
+    )
+
+    # git rev-list later_sha ^earlier_sha
+    # shows all commits reachable via later_sha but not by earlier_sha
+    #
+    # https://git-scm.com/docs/git-rev-list#_description
+    intermediate_shas = non_empty_lines(
+        popen_git(
+            cli_ctxt,
+            "rev-list",
+            later_sha,
+            "^" + earlier_sha,
+        )
+    )
+
+    for intermediate_sha in intermediate_shas:
```

Comment:
Hmmm could you clarify (preferably in code comment) why you check against all the intermediate SHAs, rather than just against `later_sha`?  🤔 I'd intuitively do the latter thing... but I'm likely missing sth